src/GridFS/StreamWrapper.php

```diff
@@ -105,16 +106,17 @@ public function rename(string $fromPath, string $toPath): bool
             throw LogicException::renamePathMismatch($fromPath, $toPath);
         }
 
-        try {
-            $this->stream_open($fromPath, 'r', 0, $openedPath);
-        } catch (FileNotFoundException $e) {
-            return false;
-        }
+        $context = $this->getContext($fromPath, 'w');
+
+        $newFilename = explode('/', $toPath, 4)[3] ?? '';
+        $count = $context['collectionWrapper']->updateFilenameForFilename($context['filename'], $newFilename);
```

Comment:
Edge case: if `$fromPath` and `$toPath` are identical, the file exists but you'll throw FileNotFoundException. That may warrant reverting to an "r" mode when selecting the context to throw eagerly.

Alternatively, you could have `updateFilenameForFilename()` return the matched count and assume that the update was successful or a NOP. The only normal scenario where the matched and modified counts wouldn't be equals is if the paths were the same, in which case you could probably NOP earlier and avoid calling `updateFilenameForFilename()` altogether (FWIW, PHP's `rename()` returns `true` in this case, but I dunno if there's some internal NOP behavior).

Reply:
Good catch. I updated the code to return the matched count, and added a test for renaming with the same filename (existing and not existing).